File: sdk/python/src/openlit/guard/schema.py
"""
Schema validation guard for structured LLM outputs.

Validates that model output is valid JSON and/or conforms to a JSON schema.
Phases: postflight only.
"""

from __future__ import annotations

import json
from typing import Any, Dict, Optional

from openlit.guard._base import Guard, GuardPhase, GuardResult


def _validate_json_schema(
    data: Any, schema: Dict[str, Any], path: str = ""
) -> Optional[str]:
    """Minimal recursive JSON schema validator (type, required, properties)."""
    schema_type = schema.get("type")
    if schema_type:
        type_map = {
            "object": dict,
            "array": list,
            "string": str,
            "number": (int, float),
            "integer": int,
            "boolean": bool,
            "null": type(None),
        }
        expected = type_map.get(schema_type)
        if expected and not isinstance(data, expected):
            return (
                f"Expected {schema_type} at {path or 'root'}, got {type(data).__name__}"
            )

    if schema_type == "object" and isinstance(data, dict):
        for field_name in schema.get("required", []):
            if field_name not in data:
                return f"Missing required field '{field_name}' at {path or 'root'}"

        properties = schema.get("properties", {})
        for prop_name, prop_schema in properties.items():
            if prop_name in data:
                err = _validate_json_schema(
                    data[prop_name], prop_schema, f"{path}.{prop_name}"
                )
                if err:
                    return err

    if schema_type == "array" and isinstance(data, list):
        items_schema = schema.get("items")
        if items_schema:
            for i, item in enumerate(data):
                err = _validate_json_schema(item, items_schema, f"{path}[{i}]")
                if err:
                    return err

    return None


class Schema(Guard):
    """
    Validates that model output is valid JSON and optionally matches a schema.

    Parameters
    ----------
    action : str
        ``"deny"`` (default) or ``"warn"``.
    json : bool
        If ``True``, just verify the output is parseable JSON.
    schema : dict, optional
        A JSON-schema-like dict to validate against.
    """

    name = "schema"
    phases = (GuardPhase.POSTFLIGHT,)

    def __init__(
        self,
        action: str = "deny",
        json_mode: bool = False,
        schema: Optional[Dict[str, Any]] = None,
        **kwargs,
    ):
        super().__init__(action=action, **kwargs)
        self._json_mode = json_mode
        self._schema = schema

    def evaluate(self, text: str) -> GuardResult:
        stripped = text.strip()

        try:
            parsed = json.loads(stripped)
        except (json.JSONDecodeError, ValueError) as exc:
            return GuardResult(
                action=self._action,
                score=1.0,
                guard_name=self.name,
                classification="invalid_json",
                explanation=f"Output is not valid JSON: {exc}",
            )

        if self._schema:
            err = _validate_json_schema(parsed, self._schema)
            if err:
                return GuardResult(
                    action=self._action,
                    score=0.9,
                    guard_name=self.name,
                    classification="schema_mismatch",
                    explanation=err,
                )

        return GuardResult(guard_name=self.name)

File: sdk/typescript/src/guard/schema.ts
/**
 * Schema validation guard for structured LLM outputs.
 *
 * Validates that model output is valid JSON and/or conforms to a JSON schema.
 *
 * Must stay in sync with: sdk/python/src/openlit/guard/schema.py
 */

import { Guard, GuardPhase, GuardResult, makeGuardResult, GuardOptions } from './base';

function validateJsonSchema(
  data: unknown,
  schema: Record<string, unknown>,
  path = '',
): string | null {
  const schemaType = schema.type as string | undefined;
  if (schemaType) {
    const typeMap: Record<string, (v: unknown) => boolean> = {
      object: (v) => typeof v === 'object' && v !== null && !Array.isArray(v),
      array: (v) => Array.isArray(v),
      string: (v) => typeof v === 'string',
      number: (v) => typeof v === 'number',
      integer: (v) => typeof v === 'number' && Number.isInteger(v),
      boolean: (v) => typeof v === 'boolean',
      null: (v) => v === null,
    };
    const checker = typeMap[schemaType];
    if (checker && !checker(data)) {
      const actual = data === null ? 'null' : Array.isArray(data) ? 'array' : typeof data;
      return `Expected ${schemaType} at ${path || 'root'}, got ${actual}`;
    }
  }

  if (schemaType === 'object' && typeof data === 'object' && data !== null && !Array.isArray(data)) {
    const dataObj = data as Record<string, unknown>;
    const required = (schema.required as string[]) || [];
    for (const fieldName of required) {
      if (!(fieldName in dataObj)) {
        return `Missing required field '${fieldName}' at ${path || 'root'}`;
      }
    }

    const properties = (schema.properties as Record<string, Record<string, unknown>>) || {};
    for (const [propName, propSchema] of Object.entries(properties)) {
      if (propName in dataObj) {
        const err = validateJsonSchema(dataObj[propName], propSchema, `${path}.${propName}`);
        if (err) return err;
      }
    }
  }

  if (schemaType === 'array' && Array.isArray(data)) {
    const itemsSchema = schema.items as Record<string, unknown> | undefined;
    if (itemsSchema) {
      for (let i = 0; i < data.length; i++) {
        const err = validateJsonSchema(data[i], itemsSchema, `${path}[${i}]`);
        if (err) return err;
      }
    }
  }

  return null;
}

export interface SchemaOptions extends GuardOptions {
  jsonMode?: boolean;
  schema?: Record<string, unknown>;
}

export class Schema extends Guard {
  readonly name = 'schema';
  readonly phases = [GuardPhase.POSTFLIGHT];

  private readonly _jsonMode: boolean;
  private readonly _schema: Record<string, unknown> | null;

  constructor(opts: SchemaOptions = {}) {
    super({ action: opts.action ?? 'deny', maxScanLength: opts.maxScanLength });
    this._jsonMode = opts.jsonMode ?? false;
    this._schema = opts.schema ?? null;
  }

  evaluate(text: string): GuardResult {
    const stripped = text.trim();

    let parsed: unknown;
    try {
      parsed = JSON.parse(stripped);
    } catch (e: unknown) {
      const message = e instanceof Error ? e.message : String(e);
      return makeGuardResult({
        action: this._action,
        score: 1.0,
        guardName: this.name,
        classification: 'invalid_json',
        explanation: `Output is not valid JSON: ${message}`,
      });
    }

    if (this._schema) {
      const err = validateJsonSchema(parsed, this._schema);
      if (err) {
        return makeGuardResult({
          action: this._action,
          score: 0.9,
          guardName: this.name,
          classification: 'schema_mismatch',
          explanation: err,
        });
      }
    }

    return makeGuardResult({ guardName: this.name });
  }
}
